File: src/main/java/com/lauriewired/GhidraMCPServer.java
package com.lauriewired;

import com.lauriewired.context.GhidraContext;
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * HTTP server wrapper that exposes GhidraAnalysisService functionality via REST
 * endpoints.
 * This class is shared between GUI and headless modes.
 */
public class GhidraMCPServer {
    private HttpServer server;
    private final GhidraAnalysisService analysisService;
    private final GhidraContext context;
    private final int port;

    public GhidraMCPServer(GhidraAnalysisService analysisService, GhidraContext context, int port) {
        this.analysisService = analysisService;
        this.context = context;
        this.port = port;
    }

    public void start() throws IOException {
        server = HttpServer.create(new InetSocketAddress(port), 0);
        setupEndpoints();
        server.setExecutor(null);
        server.start();
    }

    private void sendJsonResponse(HttpExchange exchange, Map<String, Object> response) throws IOException {
        // Simple JSON serialization - in production use a proper JSON library
        StringBuilder json = new StringBuilder();
        json.append("{");
        boolean first = true;
        for (Map.Entry<String, Object> entry : response.entrySet()) {
            if (!first)
                json.append(",");
            json.append("\"").append(entry.getKey()).append("\":");
            Object value = entry.getValue();
            if (value instanceof String) {
                json.append("\"").append(value).append("\"");
            } else if (value instanceof Boolean) {
                json.append(value);
            } else {
                json.append("\"").append(value.toString()).append("\"");
            }
            first = false;
        }
        json.append("}");

        byte[] bytes = json.toString().getBytes(StandardCharsets.UTF_8);
        exchange.getResponseHeaders().set("Content-Type", "application/json; charset=utf-8");
        exchange.sendResponseHeaders(200, bytes.length);
        try (OutputStream os = exchange.getResponseBody()) {
            os.write(bytes);
        }
    }

    public void stop() {
        if (server != null) {
            server.stop(1);
            server = null;
            System.out.println("GhidraMCP HTTP server stopped");
        }
    }

    private void setupEndpoints() {
        // Function listing endpoints
        server.createContext("/methods", exchange -> {
            Map<String, String> qparams = parseQueryParams(exchange);
            int offset = parseIntOrDefault(qparams.get("offset"), 0);
            int limit = parseIntOrDefault(qparams.get("limit"), 100);

            List<String> result = analysisService.getAllFunctionNames(offset, limit);
            sendResponse(exchange, String.join("\n", result));
        });

        server.createContext("/classes", exchange -> {
            Map<String, String> qparams = parseQueryParams(exchange);
            int offset = parseIntOrDefault(qparams.get("offset"), 0);
            int limit = parseIntOrDefault(qparams.get("limit"), 100);

            List<String> result = analysisService.getAllClassNames(offset, limit);
            sendResponse(exchange, String.join("\n", result));
        });

        // Decompilation endpoints
        server.createContext("/decompile", exchange -> {
            byte[] body = readAllBytes(exchange.getRequestBody());
            String name = new String(body, StandardCharsets.UTF_8);
            sendResponse(exchange, analysisService.decompileFunctionByName(name));
        });

        server.createContext("/decompile_function", exchange -> {
            Map<String, String> qparams = parseQueryParams(exchange);
            String address = qparams.get("address");
            sendResponse(exchange, analysisService.decompileFunctionByAddress(address));
        });

        // Function renaming endpoints
        server.createContext("/renameFunction", exchange -> {
            Map<String, String> params = parsePostParams(exchange);
            boolean success = analysisService.renameFunction(
                    params.get("oldName"), params.get("newName"));
            sendResponse(exchange, success ? "Renamed successfully" : "Rename failed");
        });

        server.createContext("/rename_function_by_address", exchange -> {
            Map<String, String> params = parsePostParams(exchange);
            boolean success = analysisService.renameFunctionByAddress(
                    params.get("function_address"), params.get("new_name"));
            sendResponse(exchange, success ? "Renamed successfully" : "Rename failed");
        });

        // Listing endpoints
        server.createContext("/segments", exchange -> {
            Map<String, String> qparams = parseQueryParams(exchange);
            int offset = parseIntOrDefault(qparams.get("offset"), 0);
            int limit = parseIntOrDefault(qparams.get("limit"), 100);

            List<String> result = analysisService.listSegments(offset, limit);
            sendResponse(exchange, String.join("\n", result));
        });

        server.createContext("/imports", exchange -> {
            Map<String, String> qparams = parseQueryParams(exchange);
            int offset = parseIntOrDefault(qparams.get("offset"), 0);
            int limit = parseIntOrDefault(qparams.get("limit"), 100);

            List<String> result = analysisService.listImports(offset, limit);
            sendResponse(exchange, String.join("\n", result));
        });

        server.createContext("/exports", exchange -> {
            Map<String, String> qparams = parseQueryParams(exchange);
            int offset = parseIntOrDefault(qparams.get("offset"), 0);
            int limit = parseIntOrDefault(qparams.get("limit"), 100);

            List<String> result = analysisService.listExports(offset, limit);
            sendResponse(exchange, String.join("\n", result));
        });

        server.createContext("/namespaces", exchange -> {
            Map<String, String> qparams = parseQueryParams(exchange);
            int offset = parseIntOrDefault(qparams.get("offset"), 0);
            int limit = parseIntOrDefault(qparams.get("limit"), 100);

            List<String> result = analysisService.listNamespaces(offset, limit);
            sendResponse(exchange, String.join("\n", result));
        });

        server.createContext("/searchFunctions", exchange -> {
            Map<String, String> qparams = parseQueryParams(exchange);
            String searchTerm = qparams.get("query");
            int offset = parseIntOrDefault(qparams.get("offset"), 0);
            int limit = parseIntOrDefault(qparams.get("limit"), 100);

            List<String> result = analysisService.searchFunctionsByName(searchTerm, offset, limit);
            sendResponse(exchange, String.join("\n", result));
        });

        // Additional endpoints that were missing
        server.createContext("/data", exchange -> {
            Map<String, String> qparams = parseQueryParams(exchange);
            int offset = parseIntOrDefault(qparams.get("offset"), 0);
            int limit = parseIntOrDefault(qparams.get("limit"), 100);

            List<String> result = analysisService.listDefinedData(offset, limit);
            sendResponse(exchange, String.join("\n", result));
        });

        server.createContext("/renameData", exchange -> {
            Map<String, String> params = parsePostParams(exchange);
            boolean success = analysisService.renameDataAtAddress(
                    params.get("address"), params.get("newName"));
            sendResponse(exchange, success ? "Data renamed successfully" : "Failed to rename data");
        });

        server.createContext("/get_function_by_address", exchange -> {
            Map<String, String> qparams = parseQueryParams(exchange);
            String address = qparams.get("address");
            sendResponse(exchange, analysisService.getFunctionByAddress(address));
        });

        server.createContext("/get_current_address", exchange -> {
            sendResponse(exchange, analysisService.getCurrentAddress());
        });

        server.createContext("/get_current_function", exchange -> {
            sendResponse(exchange, analysisService.getCurrentFunction());
        });

        server.createContext("/list_functions", exchange -> {
            Map<String, String> qparams = parseQueryParams(exchange);
            int offset = parseIntOrDefault(qparams.get("offset"), 0);
            int limit = parseIntOrDefault(qparams.get("limit"), 100);

            List<String> result = analysisService.listFunctions(offset, limit);
            sendResponse(exchange, String.join("\n", result));
        });

        server.createContext("/disassemble_function", exchange -> {
            Map<String, String> qparams = parseQueryParams(exchange);
            String address = qparams.get("address");
            sendResponse(exchange, analysisService.disassembleFunction(address));
        });

        server.createContext("/xrefs_to", exchange -> {
            Map<String, String> qparams = parseQueryParams(exchange);
            String address = qparams.get("address");
            int offset = parseIntOrDefault(qparams.get("offset"), 0);
            int limit = parseIntOrDefault(qparams.get("limit"), 100);

            List<String> result = analysisService.getXrefsTo(address, offset, limit);
            sendResponse(exchange, String.join("\n", result));
        });

        server.createContext("/xrefs_from", exchange -> {
            Map<String, String> qparams = parseQueryParams(exchange);
            String address = qparams.get("address");
            int offset = parseIntOrDefault(qparams.get("offset"), 0);
            int limit = parseIntOrDefault(qparams.get("limit"), 100);

            List<String> result = analysisService.getXrefsFrom(address, offset, limit);
            sendResponse(exchange, String.join("\n", result));
        });

        server.createContext("/function_xrefs", exchange -> {
            Map<String, String> qparams = parseQueryParams(exchange);
            String functionName = qparams.get("function_name");
            int offset = parseIntOrDefault(qparams.get("offset"), 0);
            int limit = parseIntOrDefault(qparams.get("limit"), 100);

            List<String> result = analysisService.getFunctionXrefs(functionName, offset, limit);
            sendResponse(exchange, String.join("\n", result));
        });

        server.createContext("/strings", exchange -> {
            Map<String, String> qparams = parseQueryParams(exchange);
            int offset = parseIntOrDefault(qparams.get("offset"), 0);
            int limit = parseIntOrDefault(qparams.get("limit"), 100);

            List<String> result = analysisService.getStrings(offset, limit);
            sendResponse(exchange, String.join("\n", result));
        });

        // Additional advanced endpoints
        // =============================

        server.createContext("/run_script", exchange -> {
            if ("POST".equals(exchange.getRequestMethod())) {
                Map<String, String> params = parsePostParams(exchange);
                String script = params.get("script");
                String result = analysisService.runScript(script);
                sendResponse(exchange, String.join("\n", result));
            }
        });

        // Data export endpoint
        server.createContext("/export_functions", exchange -> {
            if ("GET".equals(exchange.getRequestMethod())) {
                String result = analysisService.exportFunctions();
                Map<String, Object> response = new HashMap<>();
                response.put("functions", result);
                sendJsonResponse(exchange, response);
            } else {
                exchange.sendResponseHeaders(405, -1);
            }
        });

        // Comment endpoints
        server.createContext("/set_decompiler_comment", exchange -> {
            if ("POST".equals(exchange.getRequestMethod())) {
                Map<String, String> params = parsePostParams(exchange);
                String address = params.get("address");
                String comment = params.get("comment");

                boolean success = analysisService.setDecompilerComment(address, comment);
                Map<String, Object> response = new HashMap<>();
                response.put("success", success);
                sendJsonResponse(exchange, response);
            } else {
                exchange.sendResponseHeaders(405, -1);
            }
        });

        server.createContext("/set_disassembly_comment", exchange -> {
            if ("POST".equals(exchange.getRequestMethod())) {
                Map<String, String> params = parsePostParams(exchange);
                String address = params.get("address");
                String comment = params.get("comment");

                boolean success = analysisService.setDisassemblyComment(address, comment);
                Map<String, Object> response = new HashMap<>();
                response.put("success", success);
                sendJsonResponse(exchange, response);
            } else {
                exchange.sendResponseHeaders(405, -1);
            }
        });

        // Variable operations
        server.createContext("/rename_variable", exchange -> {
            if ("POST".equals(exchange.getRequestMethod())) {
                Map<String, String> params = parsePostParams(exchange);
                String functionName = params.get("function");
                String oldName = params.get("old_name");
                String newName = params.get("new_name");

                String result = analysisService.renameVariableInFunction(functionName, oldName, newName);
                Map<String, Object> response = new HashMap<>();
                response.put("result", result);
                sendJsonResponse(exchange, response);
            } else {
                exchange.sendResponseHeaders(405, -1);
            }
        });

        // Legacy endpoint alias
        server.createContext("/renameVariable", exchange -> {
            Map<String, String> params = parsePostParams(exchange);
            String functionName = params.get("function");
            String oldName = params.get("old_name");
            String newName = params.get("new_name");

            String result = analysisService.renameVariableInFunction(functionName, oldName, newName);
            sendResponse(exchange, result);
        });

        server.createContext("/set_local_variable_type", exchange -> {
            if ("POST".equals(exchange.getRequestMethod())) {
                Map<String, String> params = parsePostParams(exchange);
                String functionAddress = params.get("function_address");
                String variableName = params.get("variable_name");
                String newType = params.get("new_type");

                String result = analysisService.setLocalVariableType(functionAddress, variableName, newType);
                sendResponse(exchange, result);
            } else {
                exchange.sendResponseHeaders(405, -1);
            }
        });

        // Function prototype operations
        server.createContext("/set_function_prototype", exchange -> {
            Map<String, String> params = parsePostParams(exchange);
            String functionAddress = params.get("function_address");
            String prototype = params.get("prototype");

            if (functionAddress == null || prototype == null) {
                sendResponse(exchange, "Error: function_address and prototype parameters are required");
                return;
            }

            String result = analysisService.setFunctionPrototype(functionAddress, prototype);
            sendResponse(exchange, result);
        });

        // Data export operations
        server.createContext("/export_data", exchange -> {
            if ("GET".equals(exchange.getRequestMethod())) {
                Map<String, String> qparams = parseQueryParams(exchange);
                String address = qparams.get("address");
                String lengthStr = qparams.get("length");

                if (address == null || lengthStr == null) {
                    sendResponse(exchange, "Error: address and length parameters are required");
                    return;
                }

                try {
                    int length = Integer.parseInt(lengthStr);
                    String result = analysisService.exportData(address, length);
                    sendResponse(exchange, result);
                } catch (NumberFormatException e) {
                    sendResponse(exchange, "Error: Invalid length parameter: " + lengthStr);
                }
            } else {
                exchange.sendResponseHeaders(405, -1);
            }
        });

        // Type creation operations
        server.createContext("/create_type_from_c_definition", exchange -> {
            Map<String, String> params = parsePostParams(exchange);
            String cDefinition = params.get("definition");

            String result = analysisService.createTypeFromCDefinition(cDefinition);
            sendResponse(exchange, result);
        });
    }

    // Utility Methods
    // ===============

    private Map<String, String> parseQueryParams(HttpExchange exchange) {
        Map<String, String> result = new HashMap<>();
        String query = exchange.getRequestURI().getQuery();
        if (query != null) {
            String[] pairs = query.split("&");
            for (String p : pairs) {
                String[] kv = p.split("=");
                if (kv.length == 2) {
                    try {
                        String key = URLDecoder.decode(kv[0], StandardCharsets.UTF_8.name());
                        String value = URLDecoder.decode(kv[1], StandardCharsets.UTF_8.name());
                        result.put(key, value);
                    } catch (Exception e) {
                        System.err.println("Error decoding URL parameter: " + e.getMessage());
                    }
                }
            }
        }
        return result;
    }

    private Map<String, String> parsePostParams(HttpExchange exchange) throws IOException {
        byte[] body = readAllBytes(exchange.getRequestBody());
        String bodyStr = new String(body, StandardCharsets.UTF_8);
        Map<String, String> params = new HashMap<>();
        for (String pair : bodyStr.split("&")) {
            String[] kv = pair.split("=");
            if (kv.length == 2) {
                try {
                    String key = URLDecoder.decode(kv[0], StandardCharsets.UTF_8.name());
                    String value = URLDecoder.decode(kv[1], StandardCharsets.UTF_8.name());
                    params.put(key, value);
                } catch (Exception e) {
                    System.err.println("Error decoding POST parameter: " + e.getMessage());
                }
            }
        }
        return params;
    }

    private byte[] readAllBytes(InputStream inputStream) throws IOException {
        ByteArrayOutputStream buffer = new ByteArrayOutputStream();
        int nRead;
        byte[] data = new byte[1024];
        while ((nRead = inputStream.read(data, 0, data.length)) != -1) {
            buffer.write(data, 0, nRead);
        }
        buffer.flush();
        return buffer.toByteArray();
    }

    private void sendResponse(HttpExchange exchange, String response) throws IOException {
        byte[] bytes = response.getBytes(StandardCharsets.UTF_8);
        exchange.getResponseHeaders().set("Content-Type", "text/plain; charset=utf-8");
        exchange.sendResponseHeaders(200, bytes.length);
        try (OutputStream os = exchange.getResponseBody()) {
            os.write(bytes);
        }
    }

    private int parseIntOrDefault(String val, int defaultValue) {
        if (val == null || val.trim().isEmpty()) {
            return defaultValue;
        }
        try {
            return Integer.parseInt(val.trim());
        } catch (NumberFormatException e) {
            return defaultValue;
        }
    }
}

File: src/main/scala/com/lauriewired/GhidraMCPServer.scala
package com.lauriewired

import com.lauriewired.context.GhidraContext
import com.sun.net.httpserver.{HttpExchange, HttpServer}
import java.net.{InetSocketAddress, URLDecoder}
import java.nio.charset.StandardCharsets
import java.util.concurrent.{ExecutorService, Executors}

class GhidraMCPServer(analysisService: GhidraAnalysisService, context: GhidraContext):
  private var server: Option[HttpServer] = None
  private var executor: Option[ExecutorService] = None

  def start(): Unit = start(0)

  def start(port: Int): Unit =
    val s = HttpServer.create(new InetSocketAddress(port), 0)
    setupEndpoints(s)
    val exec = Executors.newCachedThreadPool(r => {
      val t = new Thread(r)
      t.setDaemon(true)
      t
    })
    s.setExecutor(exec)
    s.start()
    server = Some(s)
    executor = Some(exec)

  def getPort(): Int = server.map(_.getAddress.getPort).getOrElse(-1)

  def stop(): Unit =
    server.foreach { s =>
      s.stop(1)
      println("GhidraMCP HTTP server stopped")
    }
    executor.foreach(_.shutdownNow())
    server = None
    executor = None

  private def setupEndpoints(s: HttpServer): Unit =
    def listEndpoint(path: String, fn: (Int, Int) => Either[String, List[String]]): Unit =
      s.createContext(path, exchange => {
        val q = parseQueryParams(exchange)
        sendResponse(exchange, fn(parseInt(q.get("offset"), 0), parseInt(q.get("limit"), 100)).fold(identity, _.mkString("\n")))
      })

    listEndpoint("/methods",        analysisService.getAllFunctionNames)
    listEndpoint("/classes",        analysisService.getAllClassNames)
    listEndpoint("/segments",       analysisService.listSegments)
    listEndpoint("/imports",        analysisService.listImports)
    listEndpoint("/exports",        analysisService.listExports)
    listEndpoint("/namespaces",     analysisService.listNamespaces)
    listEndpoint("/data",           analysisService.listDefinedData)
    listEndpoint("/list_functions", analysisService.listFunctions)
    listEndpoint("/strings",        (o, l) => analysisService.getStrings(o, l))

    s.createContext("/searchFunctions", exchange => {
      val q = parseQueryParams(exchange)
      val result = analysisService.searchFunctionsByName(
        q.getOrElse("query", ""), parseInt(q.get("offset"), 0), parseInt(q.get("limit"), 100))
      sendResponse(exchange, result.fold(identity, _.mkString("\n")))
    })

    s.createContext("/xrefs_to", exchange => {
      val q = parseQueryParams(exchange)
      val result = analysisService.getXrefsTo(
        q.getOrElse("address", ""), parseInt(q.get("offset"), 0), parseInt(q.get("limit"), 100))
      sendResponse(exchange, result.fold(identity, _.mkString("\n")))
    })

    s.createContext("/xrefs_from", exchange => {
      val q = parseQueryParams(exchange)
      val result = analysisService.getXrefsFrom(
        q.getOrElse("address", ""), parseInt(q.get("offset"), 0), parseInt(q.get("limit"), 100))
      sendResponse(exchange, result.fold(identity, _.mkString("\n")))
    })

    s.createContext("/function_xrefs", exchange => {
      val q = parseQueryParams(exchange)
      val result = analysisService.getFunctionXrefs(
        q.getOrElse("function_name", ""), parseInt(q.get("offset"), 0), parseInt(q.get("limit"), 100))
      sendResponse(exchange, result.fold(identity, _.mkString("\n")))
    })

    s.createContext("/decompile", exchange => {
      val name = new String(exchange.getRequestBody.readAllBytes(), StandardCharsets.UTF_8)
      sendResponse(exchange, analysisService.decompileFunctionByName(name))
    })

    s.createContext("/decompile_function", exchange => {
      sendResponse(exchange, analysisService.decompileFunctionByAddress(
        parseQueryParams(exchange).getOrElse("address", "")))
    })

    s.createContext("/disassemble_function", exchange => {
      sendResponse(exchange, analysisService.disassembleFunction(
        parseQueryParams(exchange).getOrElse("address", "")))
    })

    s.createContext("/renameFunction", exchange => {
      val p = parsePostParams(exchange)
      sendResponse(exchange, analysisService.renameFunction(
        p.getOrElse("oldName", ""), p.getOrElse("newName", "")).fold(identity, identity))
    })

    s.createContext("/rename_function_by_address", exchange => {
      val p = parsePostParams(exchange)
      sendResponse(exchange, analysisService.renameFunctionByAddress(
        p.getOrElse("function_address", ""), p.getOrElse("new_name", "")).fold(identity, identity))
    })

    s.createContext("/create_or_update_data_item", exchange => {
      val p = parsePostParams(exchange)
      sendResponse(exchange, analysisService.createOrUpdateDataItem(
        p.getOrElse("address", ""), p.getOrElse("data_type", ""), p.getOrElse("new_name", "")).fold(identity, identity))
    })

    s.createContext("/get_function_by_address", exchange => {
      sendResponse(exchange, analysisService.getFunctionByAddress(
        parseQueryParams(exchange).getOrElse("address", "")))
    })

    s.createContext("/get_current_address", exchange =>
      sendResponse(exchange, analysisService.getCurrentAddress()))

    s.createContext("/get_current_function", exchange =>
      sendResponse(exchange, analysisService.getCurrentFunction()))

    s.createContext("/goto", exchange => {
      try
        if exchange.getRequestMethod == "POST" then
          val result = analysisService.goToTarget(parsePostParams(exchange).getOrElse("target", ""))
          sendResponse(exchange, if result != null then result else "Error: goto returned no response")
        else
          exchange.sendResponseHeaders(405, -1)
      catch case e: Exception =>
        try sendResponse(exchange, s"Error: goto endpoint failed: ${e.getMessage}")
        catch case _: Exception => System.err.println(s"Failed to send goto error response: ${e.getMessage}")
    })

    s.createContext("/status", exchange => {
      val gui = context.isGuiMode()
      val programName = context.getCurrentProgram()
        .map(p => Option(p.getDomainFile).map(_.getName).getOrElse(p.getName))
        .orNull
      sendJsonResponse(exchange, Map(
        "mode"           -> (if gui then "Headed (GUI)" else "Headless"),
        "gui"            -> gui,
        "programName"    -> programName,
        "programLoaded"  -> (programName != null),
        "currentAddress" -> analysisService.getCurrentAddress()
      ))
    })

    s.createContext("/run_script", exchange => {
      if exchange.getRequestMethod == "POST" then
        val p = parsePostParams(exchange)
        sendResponse(exchange, analysisService.runScript(p.getOrElse("name", ""), p.getOrElse("script", "")))
    })

    s.createContext("/export_functions", exchange => {
      if exchange.getRequestMethod == "GET" then
        sendJsonResponse(exchange, Map("functions" -> analysisService.exportFunctions()))
      else
        exchange.sendResponseHeaders(405, -1)
    })

    s.createContext("/set_decompiler_comment", exchange => {
      if exchange.getRequestMethod == "POST" then
        val p = parsePostParams(exchange)
        sendJsonResponse(exchange, Map("success" ->
          analysisService.setDecompilerComment(p.getOrElse("address", ""), p.getOrElse("comment", ""))))
      else
        exchange.sendResponseHeaders(405, -1)
    })

    s.createContext("/set_disassembly_comment", exchange => {
      if exchange.getRequestMethod == "POST" then
        val p = parsePostParams(exchange)
        sendJsonResponse(exchange, Map("success" ->
          analysisService.setDisassemblyComment(p.getOrElse("address", ""), p.getOrElse("comment", ""))))
      else
        exchange.sendResponseHeaders(405, -1)
    })

    s.createContext("/rename_variable", exchange => {
      if exchange.getRequestMethod == "POST" then
        val p = parsePostParams(exchange)
        sendJsonResponse(exchange, Map("result" -> analysisService.renameVariableInFunction(
          p.getOrElse("function", ""), p.getOrElse("old_name", ""), p.getOrElse("new_name", ""))))
      else
        exchange.sendResponseHeaders(405, -1)
    })

    s.createContext("/renameVariable", exchange => {
      val p = parsePostParams(exchange)
      sendResponse(exchange, analysisService.renameVariableInFunction(
        p.getOrElse("function", ""), p.getOrElse("old_name", ""), p.getOrElse("new_name", "")))
    })

    s.createContext("/set_local_variable_type", exchange => {
      if exchange.getRequestMethod == "POST" then
        val p = parsePostParams(exchange)
        sendResponse(exchange, analysisService.setLocalVariableType(
          p.getOrElse("function_address", ""), p.getOrElse("variable_name", ""), p.getOrElse("new_type", "")))
      else
        exchange.sendResponseHeaders(405, -1)
    })

    s.createContext("/set_function_prototype", exchange => {
      val p = parsePostParams(exchange)
      val addr = p.getOrElse("function_address", "")
      val proto = p.getOrElse("prototype", "")
      if addr.isEmpty || proto.isEmpty then
        sendResponse(exchange, "Error: function_address and prototype parameters are required")
      else
        sendResponse(exchange, analysisService.setFunctionPrototype(addr, proto))
    })

    s.createContext("/export_data", exchange => {
      if exchange.getRequestMethod == "GET" then
        val q = parseQueryParams(exchange)
        val addr = q.getOrElse("address", "")
        val lenStr = q.getOrElse("length", "")
        if addr.isEmpty || lenStr.isEmpty then
          sendResponse(exchange, "Error: address and length parameters are required")
        else
          scala.util.Try(lenStr.toInt).fold(
            _ => sendResponse(exchange, s"Error: Invalid length parameter: $lenStr"),
            len => sendResponse(exchange, analysisService.exportData(addr, len))
          )
      else
        exchange.sendResponseHeaders(405, -1)
    })

    s.createContext("/create_type_from_c_definition", exchange => {
      sendResponse(exchange, analysisService.createTypeFromCDefinition(
        parsePostParams(exchange).getOrElse("definition", "")))
    })

  // Response Helpers
  // ================

  private def sendResponse(exchange: HttpExchange, body: String): Unit =
    val bytes = body.getBytes(StandardCharsets.UTF_8)
    exchange.getResponseHeaders.set("Content-Type", "text/plain; charset=utf-8")
    exchange.sendResponseHeaders(200, bytes.length)
    val os = exchange.getResponseBody
    try os.write(bytes)
    finally os.close()

  private def sendJsonResponse(exchange: HttpExchange, response: Map[String, Any]): Unit =
    val json = response.map { (k, v) =>
      val encoded = v match
        case s: String  => s""""${escapeJson(s)}""""
        case b: Boolean => b.toString
        case null       => "null"
        case other      => s""""${escapeJson(other.toString)}""""
      s""""$k":$encoded"""
    }.mkString("{", ",", "}")
    val bytes = json.getBytes(StandardCharsets.UTF_8)
    exchange.getResponseHeaders.set("Content-Type", "application/json; charset=utf-8")
    exchange.sendResponseHeaders(200, bytes.length)
    val os = exchange.getResponseBody
    try os.write(bytes)
    finally os.close()

  private def escapeJson(s: String): String =
    s.replace("\\", "\\\\").replace("\"", "\\\"").replace("\n", "\\n").replace("\r", "\\r").replace("\t", "\\t")

  // Request Parsing Helpers
  // =======================

  private def parseQueryParams(exchange: HttpExchange): Map[String, String] =
    Option(exchange.getRequestURI.getQuery)
      .map(_.split("&").flatMap(decodePair).toMap)
      .getOrElse(Map.empty)

  private def parsePostParams(exchange: HttpExchange): Map[String, String] =
    new String(exchange.getRequestBody.readAllBytes(), StandardCharsets.UTF_8)
      .split("&").flatMap(decodePair).toMap

  private def decodePair(pair: String): Option[(String, String)] =
    pair.split("=", 2) match
      case Array(k, v) =>
        try Some(URLDecoder.decode(k, StandardCharsets.UTF_8) -> URLDecoder.decode(v, StandardCharsets.UTF_8))
        catch case _: Exception => None
      case _ => None

  private def parseInt(value: Option[String], default: Int): Int =
    value.flatMap(s => scala.util.Try(s.trim.toInt).toOption).getOrElse(default)
